src/contracts/core/DelegationManager.sol

```diff
@@ -224,48 +227,171 @@ contract DelegationManager is Initializable, OwnableUpgradeable, Pausable, Deleg
+        Withdrawal calldata withdrawal,
+        IERC20[] calldata tokens,
+        uint256 middlewareTimesIndex,
+        bool receiveAsTokens
+    ) external onlyWhenNotPaused(PAUSED_EXIT_WITHDRAWAL_QUEUE) {
+        bytes32 withdrawalRoot = calculateWithdrawalRoot(withdrawal);
+
+        require(
+            pendingWithdrawals[withdrawalRoot], 
+            "DelegationManager.completeQueuedAction: action is not in queue"
+        );
+
+        require(
+            slasher.canWithdraw(withdrawal.delegatedTo, withdrawal.startBlock, middlewareTimesIndex),
+            "DelegationManager.completeQueuedAction: pending action is still slashable"
+        );
+
+        require(
+            withdrawal.startBlock + withdrawalDelayBlocks <= block.number, 
+            "DelegationManager.completeQueuedAction: withdrawalDelayBlocks period has not yet passed"
+        );
+
+        require(
+            msg.sender == withdrawal.withdrawer, 
+            "DelegationManager.completeQueuedAction: only withdrawer can complete action"
+        );
+
+        if (receiveAsTokens) {
+            require(
+                tokens.length == withdrawal.strategies.length, 
+                "DelegationManager.completeQueuedAction: input length mismatch"
+            );
+        }
+
+        // Remove `withdrawalRoot` from pending roots
+        delete pendingWithdrawals[withdrawalRoot];
+
+        address currentOperator = delegatedTo[msg.sender];
+
+        // Finalize action by converting shares to tokens for each strategy, or
+        // by re-awarding shares in each strategy.
+        for (uint256 i = 0; i < withdrawal.strategies.length; ) {
+            if (receiveAsTokens) {
+                _withdrawSharesAsTokens({
+                    staker: withdrawal.staker,
+                    withdrawer: msg.sender,
+                    strategy: withdrawal.strategies[i],
+                    shares: withdrawal.shares[i],
+                    token: tokens[i]
+                });
+            } else {
+                // Award shares back to staker in StrategyManager/EigenPodManager
+                // If staker is delegated, increases shares delegated to operator
+                _addAndDelegateShares({
+                    staker: withdrawal.staker,
+                    withdrawer: msg.sender,
+                    operator: currentOperator,
+                    strategy: withdrawal.strategies[i],
+                    shares: withdrawal.shares[i]
+                });
+            }
+
+            unchecked { ++i; }
+        }
+
+        emit WithdrawalCompleted(withdrawalRoot);
+    }
+
+    /// @notice Migrates an existing queued withdrawal from the StrategyManager contract to this contract.
+    /// @dev This function is expected to be removed in the next upgrade, after all queued withdrawals have been migrated.
+    function migrateQueuedWithdrawal(IStrategyManager.DeprecatedStruct_QueuedWithdrawal memory strategyManagerWithdrawalToMigrate) external {
```

Comment:
Can this function accept an array of withdrawals to migrate? That way we don't have to go through a multicall contract for a batched withdrawal migration

Reply:
yes, confirming here that I am cool with you making this change, should just modify it so the previous withdrawal not existing is a no-op rather than causing reversion (otherwise there is a potential race condition)